src/main/java/service/CompactionService.java

```diff
@@ -1,161 +1,112 @@
-                    log.error(e);
-                    e.printStackTrace();
-                }
-            }
-
-            log.info("Sleeping for " + this.waitTime + " done:" + doneRegions);
-            Thread.sleep(this.waitTime);
+  private static Logger log = Logger.getLogger(CompactionService.class);
+
+  private Admin admin;
+  private Connection connection;
+  private CompactorConfig compactorConfig;
+  private int waitTime;
+  private TableName tableName;
+  private Map<String, RegionInfo> regionInfoMap = new HashMap<>();
+
+  public CompactionService(Connection connection, CompactorConfig config) throws IOException {
+    this.tableName = TableName.valueOf((String) config.getConfig(TABLE_NAME_KEY));
+    this.waitTime = ((int) config.getConfig(WAIT_TIME_KEY)) * 1000;
+    this.admin = connection.getAdmin();
+    this.compactorConfig = config;
+    this.connection = connection;
+    updateAllRegions();
+  }
+
+  private long getTimeAddHours(int numHours) {
+    Calendar cal = Calendar.getInstance();
+    cal.setTime(new Date());
+    cal.add(Calendar.HOUR, numHours);
+    return cal.getTime().getTime();
+  }
+
+  private void updateAllRegions() throws IOException {
+    List<RegionInfo> regionInfos = this.admin.getRegions(tableName);
+    regionInfos.stream().forEach(regionInfo -> {
+      regionInfoMap.putIfAbsent(regionInfo.getEncodedName(), regionInfo);
+    });
+  }
+
+  public void start() throws IOException, InterruptedException {
+    RegionFetcher regionFetcher = new RegionFetcher(connection, compactorConfig);
+    log.info("Starting major compaction for table " + tableName);
+
+    Set<String> compactingRegions = new HashSet<>();
+    Set<String> completedRegions = new HashSet<>();
+    int iteration = 0;
+    while (true) {
+      iteration += 1;
+      updateAllRegions();
+      log.info("Running iteration: " + iteration);
+      for (String compactingRegion : compactingRegions) {
+        CompactionState compactionState = admin.getCompactionStateForRegion(Bytes.toBytes(compactingRegion));
+        log.info("Compaction state: " + compactingRegion + " - " + compactionState);
+        if (compactionState.equals(CompactionState.MAJOR) || compactionState.equals(CompactionState.MAJOR_AND_MINOR)) {
+          log.info("In Progress compaction for: " + compactingRegion);
+        } else {
+          log.info("Compaction complete for: " + compactingRegion);
+          completedRegions.add(compactingRegion);
         }
-
-
-        log.info("compaction complete");
-    }
-
-    private void merge(Map<String, List<String>> map1, Map<String, List<String>> map2) {
-        for(Map.Entry<String, List<String>> entry : map2.entrySet()) {
-            if(map1.containsKey(entry.getKey())) {
-                map1.get(entry.getKey()).addAll(entry.getValue());
-            } else {
-                map1.put(entry.getKey(), entry.getValue());
-            }
+      }
+      compactingRegions.removeAll(completedRegions);
```

Comment:
Should we think of MAXIMUM NUMBER OF REGION in MAJOR compacting state ?
Like , if N number of Regions are in MAJOR compacting STATE, keep delaying ?

Reply:
That is there in RegionFetcher. It gives only number of regions < batch size (including already compacting). See line number: 133.